src/halmos/sevm.py

```diff
@@ -709,7 +708,7 @@ def popi(self) -> BV:
         """The stack can contain BitVecs or Bools -- this function converts Bools to BitVecs"""
 
         val = self.pop()
-        return val if type(val) is BV else BV(val, size=256)
+        return val.as_bv(size=256) if type(val) is Bool else val
```

Comment:
how can we make sure that val is either bool or bv here?  not specific to this, but overall, all the various optimizations here and there make it much harder to reason about the type consistency manually. perhaps we would need to consider incorporating actual type checking (not just type hinting), like mypy?

Reply:
yes, mypy/pyright would be amazing, but that's probably going to be a big lift to get there

there is an assert in State.push:

```
    def push(self, v: Bool | BV) -> None:
        type_v = type(v)
        assert type_v is BV and v.size == 256 or type_v is Bool
        self.stack.append(v)
```

the assert is on by default and a little bit costly at runtime, but we know for sure that State.stack only contains Bools and word-size BVs